src/cnd/cnd.ts

```diff
@@ -329,6 +329,218 @@ export interface SwapProperties {
+   */
+  // Not sure this is ideal, let's discuss and see what we can come up with
+  asset: string | { standard: "erc20"; contract_address: "0x..." };
+  /**
+   * Expiry: Expiry applying to this ledger, it can be an absolute or relative expiry, depending on the protocol and ledger
+   */
+  // We could have an object that allows the specification of the type of expiry here
+  expiry: number;
+}
+
+export enum EscrowStatus {
+  None = "NONE",
+  Initialized = "INITIALIZED",
+  Deployed = "DEPLOYED",
+  Funded = "FUNDED",
+  Redeemed = "REDEEMED",
+  Refunded = "REFUNDED",
+  IncorrectlyFunded = "INCORRECTLY_FUNDED"
+}
+
+/**
+ * The detailed description of the ledger state
+ */
+export interface LedgerState {
+  /**
+   * The address of the htlc.
+   */
+  htlc_location?: any;
+  /**
+   * The transaction ID of the deployment transaction.
+   */
+  deploy_tx?: string;
+  /**
+   * The transaction ID of the funding transaction.
+   */
+  fund_tx?: string;
+  /**
+   * The transaction ID of the redeem transaction.
+   */
+  redeem_tx?: string;
+  /**
+   * The transaction ID of the refund transaction.
+   */
+  refund_tx?: string;
+  /**
+   * The status of the escrow (HTLC or Invoice).
+   */
+  status: EscrowStatus;
+}
+
+// TODO: add `[k: string]: any;` in each section
+/**
+ * All the properties of a swap
+ */
+export interface FullSwapProperties {
+  /**
+   * The local id of the swap.
+   */
+  local_id: string;
+  /**
+   * The parameters of this swap.
+   */
+  // Decided to put everything together as now the protocols are tightly linked to the ledgers and assets
+  parameters: {
+    alpha: Parameters;
+    beta: Parameters;
+  };
+  /**
+   * The status this swap is currently in.
+   */
+  status: SwapStatus;
+  /**
+   * The data learned at creation of the swap.
+   */
+  created: {
+    /**
+     * The role in which you are participating in this swap.
+     */
+    role: "Alice" | "Bob";
+    /**
+     * The peer-id and hint address of the counterparty of this swap.
+     */
+    peer: Peer;
+    self_identities: {
+      // You have to choose between constant type (a field with a given name is always a given type) and optional fields. You cannot have both
+      /**
+       * Self Bitcoin address, if applicable
+       */
+      bitcoin?: string;
+      /**
+       * Self Ethereum address, if applicable
+       */
+      ethereum?: string;
+      /**
+       * Self Lightning node id, if applicable
+       */
+      lightning?: string;
+    };
+  };
+  /**
+   * The data learned during the communication phase.
+   *
+   * This section is only available once a swap is marked as "finalized", at the end of the communication phase.
+   * Which means it is not present when the status is "CREATED"
+   */
+  // While it could seem odd to have created/finalized section, it does simplify by reducing the number of conditional properties
+  finalized?: {
```

Comment:
I'm not sure I understand the point of splitting `created` and `finalized?`.

Reply:
Indeed, removed.